### PR TITLE
Add kmp

### PR DESCRIPTION
**Pull request by sonho00**
(Put an `X` inside the `[ ]` to denote check mark `[X]`.)

- [X] **If creating a new file :**
  - [X] added links to it in the README files ?
  - [X] included tests with it ?
  - [X] added description (overview of algorithm, time and space complexity, and possible edge case) in docstrings ?

- [ ] **if done some changes :**
  - [ ] wrote short description in the PR explaining what the changes do ?
  - [ ] Fixes #[issue number] if related to any issue

- [ ] **other**

### CODE REVIEW

File: tests/test_strings.py [hunk 1]
```diff
@@ -532,5 +533,15 @@ def test_atbash_cipher(self):
...
+class TestKnuthMorrisPratt(unittest.TestCase):
+    """[summary]
+    Test for the file knuth_morris_pratt.py
+
+    Arguments:
+        unittest {[type]} -- [description]
+    """
+    def test_knuth_morris_pratt(self):
+        self.assertEqual([0, 4], knuth_morris_pratt("abcdabc", "abc"))
+
```
Comment (keon) [inline tests/test_strings.py]: think more about possible edge cases. what if the string does not match? what if more than 1 possible matches?

Comment (sonho00) [reply tests/test_strings.py]: Sorry, the description of the algorithm was quite unkind. And I edited it now. If the string does not match, this code returns an empty list. If there are more than 1 cases, list of the start index will be returned.

Comment (keon) [reply tests/test_strings.py]: Sorry if I wasn't clear. I meant unit tests should cover those edge cases.

Comment (sonho00) [reply tests/test_strings.py]: I added test case for those cases. Can you check it?

Comment (keon) [reply tests/test_strings.py]: thanks!

**Pull request merged**